File: src/main/java/brightspark/runicmagic/spell/SpellHandler.java
package brightspark.runicmagic.spell;

import brightspark.runicmagic.RunicMagic;
import brightspark.runicmagic.handler.NetworkHandler;
import brightspark.runicmagic.message.MessageAddSpellCasting;
import brightspark.runicmagic.util.CommonUtils;
import brightspark.runicmagic.util.SpellCastData;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.server.MinecraftServer;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@Mod.EventBusSubscriber(modid = RunicMagic.MOD_ID)
public class SpellHandler
{
	private static final Map<UUID, SpellCasting> CASTS = new HashMap<>();

	@SubscribeEvent
	public static void onServerTick(TickEvent.ServerTickEvent event)
	{
		if(event.phase != TickEvent.Phase.END)
			return;

		MinecraftServer server = FMLCommonHandler.instance().getMinecraftServerInstance();
		if(server == null)
			return;
		CASTS.entrySet().removeIf(entry -> {
			EntityPlayer player = server.getPlayerList().getPlayerByUUID(entry.getKey());
			return player == null || entry.getValue().update(player.world, player, false);
		});
	}

	@SubscribeEvent
	public static void onClientTick(TickEvent.ClientTickEvent event)
	{
		if(event.phase != TickEvent.Phase.END)
			return;

		Minecraft mc = Minecraft.getMinecraft();
		boolean isSinglePlayer = mc.isSingleplayer();
		CASTS.entrySet().removeIf(entry -> {
			EntityPlayer player = mc.world.getPlayerEntityByUUID(entry.getKey());
			return player == null || entry.getValue().update(mc.world, player, isSinglePlayer);
		});
	}

	public static void addSpellCast(EntityPlayer player, Spell spell, SpellCastData data)
	{
		RunicMagic.LOG.info("Adding spell cast for spell {}", spell);
		SpellCasting casting = CASTS.compute(player.getUniqueID(), (uuid, spellCasting) -> {
			if(spellCasting != null)
				//Cancel existing spell casting
				spellCasting.getSpell().onCastCancel(player);
			//Add new spell casting
			return new SpellCasting(spell, data);
		});
		if(!player.world.isRemote && !CommonUtils.isIntegratedServer())
			NetworkHandler.network.sendToAll(new MessageAddSpellCasting(player, casting));
	}

	public static void removeSpellCast(UUID playerUuid)
	{
		CASTS.remove(playerUuid);
	}
}

File: src/main/kotlin/brightspark/runicmagic/spell/SpellHandler.kt
package brightspark.runicmagic.spell

import brightspark.runicmagic.RunicMagic
import brightspark.runicmagic.message.AddSpellCastingMessage
import brightspark.runicmagic.model.SpellCastData
import brightspark.runicmagic.util.RMUtils
import brightspark.runicmagic.util.sendToAll
import net.minecraft.client.Minecraft
import net.minecraft.entity.player.PlayerEntity
import net.minecraftforge.event.TickEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.server.ServerLifecycleHooks
import java.util.*

@Mod.EventBusSubscriber(modid = RunicMagic.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
object SpellHandler {
	private val CASTS = mutableMapOf<UUID, SpellCasting>()

	@SubscribeEvent
	fun onServerTick(event: TickEvent.ServerTickEvent) {
		if (event.phase != TickEvent.Phase.END) return
		val server = ServerLifecycleHooks.getCurrentServer()
		CASTS.entries.removeIf {
			val player = server.playerList.getPlayerByUUID(it.key)
			return@removeIf player == null || it.value.update(player.world, player, false)
		}
	}

	@SubscribeEvent
	fun onClientTick(event: TickEvent.ClientTickEvent) {
		if (event.phase != TickEvent.Phase.END) return
		val mc = Minecraft.getInstance()
		val world = mc.world ?: return
		val singlePlayer = mc.isSingleplayer
		CASTS.entries.removeIf {
			val player = world.getPlayerByUuid(it.key)
			return@removeIf player == null || it.value.update(world, player, singlePlayer)
		}
	}

	fun addSpellCast(player: PlayerEntity, spell: Spell, data: SpellCastData) {
		RunicMagic.LOG.info("Adding spell cast for $spell")
		val casting = CASTS.compute(player.uniqueID) { uuid, existing ->
			// Cancel existing spell casting
			existing?.spell?.onCastCancel(player)
			// Add new spell casting
			return@compute SpellCasting(spell, data)
		}
		// Notify all players if on a dedicated server
		if (RMUtils.isDedicatedServer)
			RunicMagic.NETWORK.sendToAll(AddSpellCastingMessage(player, casting!!))
	}

	fun removeSpellCast(playerUuid: UUID) {
		CASTS.remove(playerUuid)
	}
}
